src/pcre2_compile_class.c

```diff
@@ -2467,33 +2475,229 @@ applications.
 
 The pptr will be left pointing at the matching META_CLASS_END. */
 
-BOOL
-PRIV(compile_class_nested)(uint32_t options, uint32_t xoptions, BOOL negated,
-  uint32_t **pptr, PCRE2_UCHAR **pcode, eclass_op_info *pop_info,
-  int *errorcodeptr, compile_block *cb, PCRE2_SIZE *lengthptr)
+static BOOL
+compile_eclass_nested(eclass_context *context, BOOL negated,
+  uint32_t **pptr, PCRE2_UCHAR **pcode,
+  eclass_op_info *pop_info, PCRE2_SIZE *lengthptr)
 {
 uint32_t *ptr = *pptr;
-PCRE2_UCHAR *code = *pcode;
 #ifdef PCRE2_DEBUG
 PCRE2_UCHAR *start_code = *pcode;
 #endif
 
 /* The CLASS_IS_ECLASS bit must be set since it is a nested class. */
-PCRE2_ASSERT(*ptr == (META_CLASS | CLASS_IS_ECLASS) ||
-             *ptr == (META_CLASS_NOT | CLASS_IS_ECLASS));
+PCRE2_ASSERT(META_CODE(*ptr) == META_CLASS ||
+             META_CODE(*ptr) == META_CLASS_NOT);
 
-if (*ptr++ == (META_CLASS_NOT | CLASS_IS_ECLASS))
+if (META_CODE(*ptr) == META_CLASS_NOT)
   negated = !negated;
 
 /* Because it's a non-empty class, there must be an operand at the start. */
-if (!compile_class_binary_loose(options, xoptions, negated, &ptr, &code,
-                                pop_info, errorcodeptr, cb, lengthptr))
+(*pptr)++;
+if (!compile_class_binary_loose(context, negated, pptr, pcode,
+                                pop_info, lengthptr))
   return FALSE;
 
-PCRE2_ASSERT(*ptr == META_CLASS_END);
-PCRE2_ASSERT(lengthptr == NULL || code == start_code);
+PCRE2_ASSERT(**pptr == META_CLASS_END);
+PCRE2_ASSERT(lengthptr == NULL || *pcode == start_code);
+return TRUE;
+}
+
+BOOL
+PRIV(compile_class_nested)(uint32_t options, uint32_t xoptions,
+  uint32_t **pptr, PCRE2_UCHAR **pcode, int *errorcodeptr,
+  compile_block *cb, PCRE2_SIZE *lengthptr)
+{
+eclass_context context;
+eclass_op_info op_info;
+PCRE2_SIZE previous_length = (lengthptr != NULL)? *lengthptr : 0;
+PCRE2_UCHAR *code = *pcode;
+PCRE2_UCHAR *previous;
+BOOL allbitsone = TRUE;
+
+context.eclass_start = *pptr;
```

Comment:
Storing this bit in the META code seems very dangerous. It carries the risk that different branches will be taken on the two passes through the code.

The flag you want should just be a BOOL on the eclass_context, so it's got a fresh state on each pass.

Reply:
That was the whole point of the patch. Although I realized you inject the bitmap at the end, so caching this value is likely unnecessary.